File: lib/features/additional_info/domain/entities/additional_info.dart
enum Gender { male, female, other }

class AdditionalInfo {
  final String? gender;
  final DateTime? birthDate;
  final int? age;
  final double? weight; // in kg
  final double? height; // in cm
  final String? activityLevel;
  final String? weightGoal;
  final String? workoutFrequency;
  final double? targetWeight; // in kg
  final double? weightLossSpeed; // in kg per week
  final String? diet;
  final String? accomplishment;
  final String? referralCode;

  const AdditionalInfo({
    this.gender,
    this.birthDate,
    this.age,
    this.weight,
    this.height,
    this.activityLevel,
    this.weightGoal,
    this.workoutFrequency,
    this.targetWeight,
    this.weightLossSpeed,
    this.diet,
    this.accomplishment,
    this.referralCode,
  });

  AdditionalInfo copyWith({
    String? gender,
    DateTime? birthDate,
    int? age,
    double? weight,
    double? height,
    String? activityLevel,
    String? weightGoal,
    String? workoutFrequency,
    double? targetWeight,
    double? weightLossSpeed,
    String? diet,
    String? accomplishment,
    String? referralCode,
  }) {
    return AdditionalInfo(
      gender: gender ?? this.gender,
      birthDate: birthDate ?? this.birthDate,
      age: age ?? this.age,
      weight: weight ?? this.weight,
      height: height ?? this.height,
      activityLevel: activityLevel ?? this.activityLevel,
      weightGoal: weightGoal ?? this.weightGoal,
      workoutFrequency: workoutFrequency ?? this.workoutFrequency,
      targetWeight: targetWeight ?? this.targetWeight,
      weightLossSpeed: weightLossSpeed ?? this.weightLossSpeed,
      diet: diet ?? this.diet,
      accomplishment: accomplishment ?? this.accomplishment,
      referralCode: referralCode ?? this.referralCode,
    );
  }

  Map<String, dynamic> toJson() => {
        'gender': gender,
        'birthDate': birthDate?.toIso8601String(),
        'age': age,
        'weight': weight,
        'height': height,
        'activityLevel': activityLevel,
        'weightGoal': weightGoal,
        'workoutFrequency': workoutFrequency,
        'targetWeight': targetWeight,
        'weightLossSpeed': weightLossSpeed,
        'diet': diet,
        'accomplishment': accomplishment,
        'referralCode': referralCode,
      };

  factory AdditionalInfo.fromJson(Map<String, dynamic> json) => AdditionalInfo(
        gender: json['gender'],
        birthDate: json['birthDate'] != null
            ? DateTime.parse(json['birthDate'])
            : null,
        age: json['age'],
        weight: json['weight']?.toDouble(),
        height: json['height']?.toDouble(),
        activityLevel: json['activityLevel'],
        weightGoal: json['weightGoal'],
        workoutFrequency: json['workoutFrequency'],
        targetWeight: json['targetWeight']?.toDouble(),
        weightLossSpeed: json['weightLossSpeed']?.toDouble(),
        diet: json['diet'],
        accomplishment: json['accomplishment'],
        referralCode: json['referralCode'],
      );

  bool get isComplete {
    final bool baseComplete = gender != null &&
        birthDate != null &&
        weight != null &&
        height != null &&
        activityLevel != null &&
        weightGoal != null &&
        workoutFrequency != null &&
        targetWeight != null &&
        diet != null &&
        accomplishment != null;

    // Require speed for both lose and gain goals
    if (weightGoal == 'lose_weight' || weightGoal == 'gain_weight') {
      return baseComplete && weightLossSpeed != null;
    }
    return baseComplete;
  }

  @override
  String toString() =>
      'AdditionalInfo(gender: $gender, birthDate: $birthDate, age: $age, weight: $weight, height: $height, activityLevel: $activityLevel, weightGoal: $weightGoal, workoutFrequency: $workoutFrequency, targetWeight: $targetWeight, weightLossSpeed: $weightLossSpeed, diet: $diet, accomplishment: $accomplishment, referralCode: $referralCode)';

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is AdditionalInfo &&
        other.gender == gender &&
        other.birthDate == birthDate &&
        other.age == age &&
        other.weight == weight &&
        other.height == height &&
        other.activityLevel == activityLevel &&
        other.weightGoal == weightGoal &&
        other.workoutFrequency == workoutFrequency &&
        other.targetWeight == targetWeight &&
        other.weightLossSpeed == weightLossSpeed &&
        other.diet == diet &&
        other.accomplishment == accomplishment &&
        other.referralCode == referralCode;
  }

  @override
  int get hashCode =>
      gender.hashCode ^
      birthDate.hashCode ^
      age.hashCode ^
      weight.hashCode ^
      height.hashCode ^
      activityLevel.hashCode ^
      weightGoal.hashCode ^
      workoutFrequency.hashCode ^
      targetWeight.hashCode ^
      weightLossSpeed.hashCode ^
      diet.hashCode ^
      accomplishment.hashCode ^
      referralCode.hashCode;
}

File: server/src/models/AdditionalInfo.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IAdditionalInfo extends Document {
    userId: mongoose.Types.ObjectId;
    gender?: string;
    age?: number;
    weight?: number; // in kg
    height?: number; // in cm
    activityLevel?: string;
    weightGoal?: string;
    workoutFrequency?: string;
    weightLossSpeed?: number; // in kg per week
    createdAt: Date;
    updatedAt: Date;
}

const additionalInfoSchema = new Schema<IAdditionalInfo>(
    {
        userId: {
            type: Schema.Types.ObjectId,
            ref: 'User',
            required: true,
            unique: true,
        },
        gender: {
            type: String,
            enum: ['male', 'female', 'other'],
            required: false,
        },
        age: {
            type: Number,
            min: [1, 'Age must be at least 1'],
            max: [120, 'Age cannot exceed 120'],
            required: false,
        },
        weight: {
            type: Number,
            min: [20, 'Weight must be at least 20 kg'],
            max: [300, 'Weight cannot exceed 300 kg'],
            required: false,
        },
        height: {
            type: Number,
            min: [100, 'Height must be at least 100 cm'],
            max: [250, 'Height cannot exceed 250 cm'],
            required: false,
        },
        activityLevel: {
            type: String,
            enum: ['sedentary', 'lightly_active', 'moderately_active', 'very_active'],
            required: false,
        },
        weightGoal: {
            type: String,
            enum: ['lose_weight', 'maintain_weight', 'gain_weight'],
            required: false,
        },
        workoutFrequency: {
            type: String,
            enum: ['0-2', '3-5', '6+'],
            required: false,
        },
        weightLossSpeed: {
            type: Number,
            min: [0.1, 'Weight loss speed must be at least 0.1 kg per week'],
            max: [2.0, 'Weight loss speed cannot exceed 2.0 kg per week'],
            required: false,
        },
    },
    {
        timestamps: true,
    }
);

// Indexes for better query performance
additionalInfoSchema.index({ userId: 1 });

const AdditionalInfo = mongoose.model<IAdditionalInfo>('AdditionalInfo', additionalInfoSchema);

export default AdditionalInfo; 